### PR TITLE
feat: make `appState.selectedElementIds` more stable

### PR DESCRIPTION
**Pull request by dwelle**
This PR optimizes `appState.selectedElementIds` so it retains its reference identity as much as possible (to optimize memoization), by:

1. making the object immutable
2. updating the object only if properties changed (we often set `selectedElementIds` to `{}` which can be a noop. Similarly when recreating the object without changing the ids within).
3. fix cases such as setting the `selection` element `id`

For this purpose we're now piping `selectedElementIds` updates through `makeNextSelectedElementIds()` helper that compares first. We could theoretically monkeypatch the `App.setState()` function to do this automatically, but I'm not certain we don't depend on intermediate `selectedElementIds` values during updates (though it's a venue to be investigated to simplify the codebase).

~~The function can be further optimized by skipping `undefined/false` pairs (when an id to `false` where it previously wasn't set all, or vice versa), and comparing the size of both objects only when necessary.~~ DONE

### CODE REVIEW

File: src/components/App.tsx [hunk 1]
```diff
@@ -5524,14 +5564,9 @@ class App extends React.Component<AppProps, AppState> {
...
                 this.scene.getNonDeletedElements(),
+                prevState,
               ),
             );
-          } else {
-            this.setState({
-              selectedElementIds: {},
-              selectedGroupIds: {},
-              editingGroupId: null,
-            });
```
Comment (dwelle) [inline src/components/App.tsx:5534]: This (alongside the change below on 5555) feels like it should break something, but I couldn't find it.

What I'm solving by removing this is to not reset `selectedElementIds` to `{}` and right after set it to element ids within active box selection. This way it was always renewing the reference and thus during `pointermove` when selecting we'd not be able to memoize on the object.

File: src/scene/selection.ts [hunk 2]
```diff
@@ -192,16 +192,38 @@ export const getTargetElements = (
...
+  }
+
+  if (!changed) {
+    const prevIds = Object.keys(prevState.selectedElementIds);
+    if (nextIds.length === prevIds.length) {
+      return prevState.selectedElementIds;
+    }
+
+    for (const id of prevIds) {
+      const prevVal = !!prevState.selectedElementIds[id];
```
Comment (Mrazator) [inline src/scene/selection.ts:223]: Is the double negation really needed here?

File: src/scene/selection.ts [hunk 3]
```diff
@@ -192,16 +192,38 @@ export const getTargetElements = (
...
+    }
+  }
+
+  if (!changed) {
+    const prevIds = Object.keys(prevState.selectedElementIds);
+    if (nextIds.length === prevIds.length) {
+      return prevState.selectedElementIds;
+    }
+
+    for (const id of prevIds) {
```
Comment (Mrazator) [inline src/scene/selection.ts:222]: Is this optimisation worth iterating through all IDs again (in a big enough scene)?

File: src/scene/selection.ts [hunk 4]
```diff
@@ -88,11 +88,41 @@ export const getElementsWithinSelection = (
...
+// FIXME move this into the editor instance to keep utility methods stateless
+export const isSomeElementSelected = (function () {
+  let lastElements: readonly NonDeletedExcalidrawElement[] | null = null;
+  let lastSelectedElementIds: AppState["selectedElementIds"] | null = null;
+  let isSelected: boolean | null = null;
+
+  const ret = (
+    elements: readonly NonDeletedExcalidrawElement[],
+    appState: Pick<AppState, "selectedElementIds">,
+  ): boolean => {
```
Comment (Mrazator) [inline src/scene/selection.ts:100]: Would be nice to do this caching separatly, I've already rewritten parts of it

File: src/groups.ts [hunk 5]
```diff
@@ -91,6 +100,11 @@ export const selectGroupsForSelectedElements = (
     }
   }
 
+  nextAppState.selectedElementIds = makeNextSelectedElementIds(
+    nextAppState.selectedElementIds,
```
Comment (Mrazator) [inline src/groups.ts:104]: Does this actually improve anything? This function is being called on every mouse down, so each time `makeNextSelectedElementIds` would be recalculated.  I've rewritten this whole function due to `selectGroup` being a bottleneck, IMO this would be better off with caching (which I already have). Would prefer to leave it for a separate PR.

File: src/actions/actionGroup.tsx [hunk 6]
```diff
@@ -232,9 +233,18 @@ export const actionUngroup = register({
     });
 
     // remove binded text elements from selection
-    boundTextElementIds.forEach(
-      (id) => (updateAppState.selectedElementIds[id] = false),
+    updateAppState.selectedElementIds = Object.fromEntries(
+      Object.entries(updateAppState.selectedElementIds).reduce(
+        (acc: [ExcalidrawElement["id"], true][], [id, selected]) => {
+          if (selected && !boundTextElementIds.includes(id)) {
+            acc.push([id, true]);
```
Comment (Mrazator) [inline src/actions/actionGroup.tsx:240]: Is it necessary to use  `Object.fromEntries(Object.entries` if could be already an object?

File: src/actions/actionFinalize.tsx [hunk 7]
```diff
@@ -125,13 +125,6 @@ export const actionFinalize = register({
...
         );
       }
-
-      if (
-        !appState.activeTool.locked &&
-        appState.activeTool.type !== "freedraw"
-      ) {
-        appState.selectedElementIds[multiPointElement.id] = true;
-      }
     }
```
Comment (Mrazator) [inline src/actions/actionFinalize.tsx:128]: Whas this removal intentional?

File: src/components/App.tsx [hunk 8]
```diff
@@ -2130,8 +2145,9 @@ class App extends React.Component<AppProps, AppState> {
     }
 
     this.setState({
-      selectedElementIds: Object.fromEntries(
-        textElements.map((el) => [el.id, true]),
+      selectedElementIds: makeNextSelectedElementIds(
+        Object.fromEntries(textElements.map((el) => [el.id, true])),
```
Comment (Mrazator) [inline src/components/App.tsx:2149]: Single reduce is more performant than first creating array and then transforming it into object

File: src/scene/selection.ts (hunk 2 above)
Comment (dwelle) [reply src/scene/selection.ts:223]: right, we're only checking for truthy values, so not needed

File: src/actions/actionBoundText.tsx [hunk 9]
```diff
@@ -211,7 +212,7 @@ export const actionWrapTextInContainer = register({
       appState,
     );
     let updatedElements: readonly ExcalidrawElement[] = elements.slice();
-    const containerIds: AppState["selectedElementIds"] = {};
+    const containerIds: Mutable<AppState["selectedElementIds"]> = {};
```
Comment (Mrazator) [inline src/actions/actionBoundText.tsx:215]: Could we add this also elsewhere?

File: src/actions/actionFinalize.tsx (hunk 7 above)
Comment (dwelle) [reply src/actions/actionFinalize.tsx:128]: I should have commented, but if you check the return value, we're doing the same thing there

File: src/actions/actionBoundText.tsx (hunk 9 above)
Comment (dwelle) [reply src/actions/actionBoundText.tsx:215]: you mean this annotation instead of adhoc `{[id: string]: true}`?

Comment (Mrazator) [reply src/actions/actionBoundText.tsx:215]: seems reasonable for the temp object

File: src/components/App.tsx (hunk 8 above)
Comment (dwelle) [reply src/components/App.tsx:2149]: this is not a hot path so it's fine

File: src/actions/actionGroup.tsx (hunk 6 above)
Comment (dwelle) [reply src/actions/actionGroup.tsx:240]: yeah, I wasn't thinking

File: src/groups.ts (hunk 5 above)
Comment (dwelle) [reply src/groups.ts:104]: irrespective of being cached or not, we want the cache-miss to not update the `selectedElementIds` unnecessarily. Generally, whenever we update the `selectedElementIds`, we should use this helper, without giving it much thought. It shouldn't hurt.

**Review by Mrazator — APPROVED** [review]

**Comment by dwelle** [comment]
❤️
